File: tuple-types-test/tuple-types-test.cpp
#include "tuple-types-test.h"

using namespace js;

const std::tuple<js::number, js::number> coordinates = js::array<js::number>{js::number(10), js::number(20)};

const std::tuple<js::number, js::number, js::number> point3D = js::array<js::number>{js::number(1), js::number(2), js::number(3)};

const std::tuple<js::string, js::number, bool> mixedTuple = js::array<js::string>{"hello"_S, js::number(42), true};

const NamedCoordinate namedPoint = js::array<js::number>{js::number(5), js::number(10)};

const std::tuple<js::string, std::optional<js::number>> optionalTuple = js::array<js::string>{"required"_S};

const std::tuple<js::string, std::optional<js::number>> optionalWithValue = js::array<js::string>{"with"_S, js::number(123)};

const std::tuple<js::string> tupleWithRest = js::array<js::string>{"first"_S, js::number(1), js::number(2), js::number(3), js::number(4), js::number(5)};

const std::tuple<bool, js::string> anotherRest = js::array<bool>{true, "hello"_S, "world"_S, "!"_S};

std::tuple<js::number, js::number> getCoordinates() {
    return js::array<js::number>{js::number(100), js::number(200)};
}

js::number processPoint(std::tuple<js::number, js::number> point) {
    auto _temp9803 = point;
    const auto x = _temp9803[0];
    const auto y = _temp9803[1];
    return (x + y);
}

auto _temp9771 = coordinates;
const auto x1 = _temp9771[0];
const auto y1 = _temp9771[1];

auto _temp8009 = mixedTuple;
const auto _str = _temp8009[0];
const auto _num = _temp8009[1];
const auto _bool = _temp8009[2];

const js::any _firstCoord = coordinates[js::number(0)];

const js::any _secondCoord = coordinates[js::number(1)];

js::array<std::tuple<js::number, js::number>> _points = js::array<js::any>{js::array<js::number>{js::number(0), js::number(0)}, js::array<js::number>{js::number(1), js::number(1)}, js::array<js::number>{js::number(2), js::number(4)}, js::array<js::number>{js::number(3), js::number(9)}};

const std::tuple<js::string, std::tuple<js::number, bool>> _nestedTuple = js::array<js::string>{"outer"_S, js::array<js::number>{js::number(42), false}};

const std::tuple<std::tuple<js::string, js::number>, std::tuple<bool, js::string>> _tripleNested = js::array<js::any>{js::array<js::string>{"a"_S, js::number(1)}, js::array<bool>{true, "b"_S}};

template<typename... Args>
void processRestTuple(Args... args) {
    auto args_array = js::array<js::any>{args...};
    auto _temp4366 = args_array;
    const auto str = _temp4366[0];
    const auto num = _temp4366[1];
    const auto bools = _temp4366.slice(2);
    js::console.log("String:"_S, str);
    js::console.log("Number:"_S, num);
    js::console.log("Booleans:"_S, bools);
}

auto _temp7021 = getCoordinates();
const auto resultX = _temp7021[0];
const auto resultY = _temp7021[1];

// Entry point
void Main() {
    js::console.log("Destructured coordinates:"_S, x1, y1);
    js::console.log("Coordinates:"_S, coordinates);
    js::console.log("Point 3D:"_S, point3D);
    js::console.log("Mixed tuple:"_S, mixedTuple);
    js::console.log("Sum of coordinates:"_S, processPoint(coordinates));
    js::console.log("Result coordinates:"_S, resultX, resultY);
    processRestTuple("test"_S, js::number(42), true, false, true);
}

int main(int /*argc*/, char** /*argv*/) {
    Main();
    return 0;
}

File: tests/integration/tuple-types-test.ts
// Test for Tuple Types support

// Basic tuple types
const coordinates: [number, number] = [10, 20];
const point3D: [number, number, number] = [1.0, 2.0, 3.0];

// Mixed type tuple
const mixedTuple: [string, number, boolean] = ["hello", 42, true];

// Named tuple members (TypeScript 4.0+)
type NamedCoordinate = [x: number, y: number];
const namedPoint: NamedCoordinate = [5, 10];

// Optional elements in tuples
const optionalTuple: [string, number?] = ["required"];
const optionalWithValue: [string, number?] = ["with", 123];

// Rest elements in tuples
const tupleWithRest: [string, ...number[]] = ["first", 1, 2, 3, 4, 5];
const anotherRest: [boolean, string, ...string[]] = [true, "hello", "world", "!"];

// Function returning tuple
function getCoordinates(): [number, number] {
  return [100, 200];
}

// Function accepting tuple parameter
function processPoint(point: [number, number]): number {
  const [x, y] = point;
  return x + y;
}

// Destructuring tuples
const [x1, y1] = coordinates;
console.log("Destructured coordinates:", x1, y1);

const [_str, _num, _bool] = mixedTuple;

// Accessing tuple elements
const _firstCoord = coordinates[0];
const _secondCoord = coordinates[1];

// Tuple in array
const _points: [number, number][] = [
  [0, 0],
  [1, 1],
  [2, 4],
  [3, 9],
];

// Complex nested tuples
const _nestedTuple: [string, [number, boolean]] = ["outer", [42, false]];
const _tripleNested: [[string, number], [boolean, string]] = [["a", 1], [true, "b"]];

// Function with rest tuple parameters
function processRestTuple(...args: [string, number, ...boolean[]]): void {
  const [str, num, ...bools] = args;
  console.log("String:", str);
  console.log("Number:", num);
  console.log("Booleans:", bools);
}

// Using tuples
console.log("Coordinates:", coordinates);
console.log("Point 3D:", point3D);
console.log("Mixed tuple:", mixedTuple);
console.log("Sum of coordinates:", processPoint(coordinates));

const [resultX, resultY] = getCoordinates();
console.log("Result coordinates:", resultX, resultY);

// Test rest tuple
processRestTuple("test", 42, true, false, true);
